spec/controllers/playlists_controller_spec.rb

```diff
@@ -146,6 +146,66 @@
     end
   end
 
+  describe "update" do
+    let(:user) { users(:jamie_kiesl) }
+
+    before do
+      login(user)
+    end
+
+    it "should update published_at" do
+      expect(playlists(:jamie_kiesl_loves).published_at).to be_nil
+      expect(playlists(:jamie_kiesl_loves).private).to be(true)
+
+      put :update, params: {
+        id: playlists(:jamie_kiesl_loves).id,
+        permalink: 'jamie-loves',
+        user_id: user.login,
+        playlist: {
+          private: false
+        }}
+      expect(playlists(:jamie_kiesl_loves).reload.published_at).not_to be_nil
+    end
+
+    it "should not publish if playlist has less than 2 tracks" do
+      expect(playlists(:jamie_kiesl_playlist_with_soft_deleted_tracks).tracks.count).to eq(1)
+
+      put :update, params: {
+        id: playlists(:jamie_kiesl_playlist_with_soft_deleted_tracks).id,
+        permalink: 'jamie-playlist-with-soft-delete',
+        user_id: user.login,
+        playlist: {
+          private: false
+        }}
+      expect(playlists(:jamie_kiesl_playlist_with_soft_deleted_tracks).reload.published_at).to be_nil
+    end
+
+    it "should set is_mix to true if is_favorite?" do
+      put :update, params: {
+        id: playlists(:jamie_kiesl_loves).id,
+        permalink: 'jamie-loves',
+        user_id: user.login,
+        playlist: {
+          is_favorite: true
+        }}
+      expect(playlists(:jamie_kiesl_loves).reload.is_mix).to eq(true)
+    end
+
+    it "should set is_mix to true if there is a mix of tracks" do
+      expect(playlists(:jamie_kiesl_loves).tracks.count).to eq(4)
+      expect(playlists(:jamie_kiesl_loves).is_mix).to eq(nil)
+      put :update, params: {
+        id: playlists(:jamie_kiesl_loves).id,
+        permalink: 'jamie-loves',
+        user_id: user.login,
+        playlist: {
+          is_favorite: false
```

Comment:
Hmmm, I see what you are saying, but `add_track` doesn't seem to be actually triggering an `update` action. Which leads me to believe that we might be making some unnecessary calls (in that method) during each actual update action.

Is it possible you are thinking of the good old days when add_track wasn't async? I was playing around `playlist/edit` page and you don't actually have to click `Save` to add a track with `+`. 

What do you think about making a ticket to clean up `set_mix_or_album` (make it an explicit action)? I could add it here, but it'd blow this PR up a bit.

Reply:
Happy to clean up mixes here if it makes sense, otherwise feel free to back out of those mix specs and changes. Track add/removing was always async. Not sure what happened here, let's should just make it nice...

One option is to use `before_add` https://guides.rubyonrails.org/v5.1/association_basics.html#association-callbacks — model callbacks might feel too magical, so we can also call `@playlist.set_mix_or_album` explicitly from the controller, since tracks are added and removed by 2 actions only.

Reply:
Few things I've tried that didn't work when trying to write a spec for this:
- When adding a first favorites track, new playlist is created, but it doesn't trigger `before_update` action
```
  Playlist Create (14.4ms)  INSERT INTO `playlists` (`title`, `user_id`, `created_at`, `updated_at`, `permalink`, `is_favorite`) VALUES ('innatewonderer\'s favorite tracks', 34856, '2019-11-05 05:16:26', '2019-11-05 05:16:26', 'innatewonderers-favorite-tracks', TRUE)
```

- when I favor a track and it makes it to my favorites playlist, but it triggers `"/toggle_favorite?asset_id=111783"`
```
Started GET "/toggle_favorite?asset_id=111783" for ::1 at 2019-11-05 00:05:21 -0500
Processing by UsersController#toggle_favorite as JS
```
The interesting thing here is that it DOES trigger `Playlist Update All (0.5ms)...`.
```
  Playlist Update All (0.5ms)  UPDATE `playlists` SET `playlists`.`tracks_count` = COALESCE(`playlists`.`tracks_count`, 0) + 1, `playlists`.`updated_at` = '2019-11-05 05:21:57' WHERE `playlists`.`id` = 20219
```
However, it skips all the callbacks. My current version is that counter_cache functionality skips callbacks. I didn't find anything in [the docs](https://guides.rubyonrails.org/association_basics.html#options-for-belongs-to-counter-cache). Found this in the [rails codebase](https://github.com/rails/rails/blob/a592e87a278cec10569d854dd5ece22179471bc0/activerecord/lib/active_record/counter_cache.rb#L65):
> It simply does a direct SQL update for the record
> with the given ID, altering the given hash of counters by the amount
> given by the corresponding value

- In order to be able to trigger `add_track`, I click the `+` on playlists/edit but the only options there are user’s own tracks (so we can’t really add other tracks that are is_favorite

Proposal for `is_mix` flag:
- Move `self.is_mix = true if is_favorite?` directly to when a new favorite's playlist is created. Then my prooving spec can cover creation of new favorites playlist and confirm that the flag is explicitely set.
- Remove `self.is_mix = true if tracks.present? && tracks.count > tracks.where('user_id != ?', user&.id).count` completely. It's a check that's performed on every `playlist.update` and not necessary
- Remove `set_mix_or_album` callaback

Thoughts?

Reply:
> the only options there are user’s own tracks (so we can’t really add other tracks that are is_favorite

The new playlist edit UI doesn't yet expose the other options, but we will be rolling those out (and have it designed) so we should keep the mix/album distinction on the backend.

> It's a check that's performed on every `playlist.update` and not necessary

You are right that a favorites playlist is always a mix. However, on normal track add, we do need to check if a playlist contains tracks from a user other than the user making the playlist and set `is_mix` to be true. +1 to removing the callback though! We can do that in `add_track`/`remove_track` (which ideally get moved into a new tracks_controller as `create`/`destroy` as those actions haven't been updated since REST was a thing!!).

Reply:
> add_track/remove_track (which ideally get moved into a new tracks_controller as create/destroy 
Another ticket #894